lib/graphql/internal_representation/rewrite.rb

```diff
@@ -137,9 +141,9 @@ def validate(context)
         # can be shared between its usages.
         context.on_dependency_resolve do |defn_ast_node, spread_ast_nodes, frag_ast_node|
           frag_name = frag_ast_node.name
-          fragment_node = fragment_definitions[frag_name]
+          fragment_node = @definitions[frag_name]
 
-          if fragment_node
+          if fragment_node && fragment_node.ast_node.is_a?(Nodes::FragmentDefinition)
```

Comment:
Crap. I think you're right. I've been assuming **definition** names where a unique set up until this point. I can't find anything in the GraphQL spec about this. It only mentions operation uniqueness and fragment definition uniqueness.

I will change this back.

It this is all true, I think theres probably a few graphql-ruby implementation bugs around this issue. I just tested on the github.com API and ran into this:

``` graphql
query Foo {
  viewer {
    ...Foo
  }
}

fragment Foo on User {
  login
}
```

``` json
{
  "data": null,
  "errors": [
    {
      "message": "Fragment Foo contains an infinite loop",
      "locations": [
        {
          "line": 7,
          "column": 1
        }
      ]
    }
  ]
}
```